### DISCUSSION TITLE
MJ-ST01 with TuyaMCU, LED indicator status light control

### DISCUSSION THREAD

**Discussion by badhras** [arendst/Tasmota#17703] [Q&A] [answered]
**BLUF update with answer: MJ-ST01 with pre-flashed Tasmota have LED indicators hooked up to a TuyaMCU. Impossible to control from Tasmota.**

I just purchased a bunch of [Martin Jerry 3-way smart switches](https://smile.amazon.com/gp/product/B0BJ939TC2/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&th=1) that are pre-flashed with Tasmota. The Tasmota part essentially works, and I've successfully configured them to integrate with HomeAssistant. However, I'm unable to configure the behavior so that when the lights are OFF, then RED LED is ON. 

I've tried using the template info from https://templates.blakadder.com/martin-jerry-3-way-ST01.html and https://github.com/arendst/Tasmota/blob/development/TEMPLATES.md and when I activate those templates, I can't even get the tasmota UI's toggle button to work. When I inquired with Martin Jerry, they noted that the 3-way switches I purchased use Tuya. I am able to use TuyaSend to turn the switch ON/OFF, but that's about it. I haven't found anything specific for controlling the LED indicator/status lights. I've asked for documentation about how they mapped the LED indicators to Tuya dpid values, but I haven't heard anything from them yet. 

I've done a little bit of web searching for dpids, and I haven't seen anything for controlling those indicator lights. Is this something that Tuya devices normally prohibit? I'm curious if the community has done anything like this with Tuya devices.

**Comment by barbudor** [comment, marked as answer]
So when the device came, it was working with Tasmota fine except for the Led problem?
May be you can start from here and tell us which version of Tasmota is used and what is the template that works.

**Reply by badhras** [reply]
For any avoidance of doubt, I purchased these [Tasmota-flashed MJ-ST01 products](https://smile.amazon.com/gp/product/B0BJ939TC2/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&th=1).

In a support chat with Martin Jerry, they clarified the following:

1. The TASMOTA version is using MCU, thus there is no GPIO info of the leds;
2. The non-TASMOTA version is using GPIO to control it.

[Information] from unmodified 3-way switch
```
Program Version:          12.1.0(ST01 by MJ)
Build Date & Time:        2022-09-07T07:33:18
Core/SDK Version:         2_7_4_9/2.2.2-dev(38a443e)
```

[Information] from modified 3-way switch
```
Program Version:          12.3.1(tasmota)
Build Date & Time:        2022-12-16T10:28:17
Core/SDK Version:         2_7_4_9/2.2.2-dev(38a443e)
```

[Configure Other] is the same for both:
```
Template:                 {"NAME":"3-way Switch","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}
Activate:                 unchecked
```

[Configure Module] is the same for both:
```
Module Type:              Tuya MCU (54)
GPIO1:                    Tuya Tx
GPIO3:                    Tuya Rx
All other GPIO:           None
```

[Configure Template] is the same for both:
```
Based On:                 Generic (18)
All GPIO:                 User
```

Exhibited behaviors are the same for both:

- When lights are on, the GREEN LED is on.
- When lights are off, the RED LED is off.

Desired change:

- When lights are off, the RED LED is on.

If [Configure Other]/Activate is checked with this template (default)
`{"NAME":"3-way Switch","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}`
then the Main page's Toggle button disappears.

If [Configure Other]/Activate is checked with this template (from [blakadder](https://templates.blakadder.com/martin-jerry-3-way-ST01.html) or [tasmota-github](https://github.com/TheEebb/tasmota-tuyamcu-fix/blob/master/TEMPLATES.md))
`{"NAME":"3-way Switch","GPIO":[255,255,255,255,52,53,0,0,21,9,157,255,0],"FLAG":0,"BASE":18}`
then the Main page's Toggle button is present, but clicking it causes no change to the switch/lights.
Using "switchmode1 5" in the [Console] doesn't change the behavior.

**Reply by barbudor** [reply]
Your original template is the one that counts, especially as it is the one provided by the vendor.
It is based only TuyaMCU and the leds are not under control of the ESP through GPIOs
So there is no point in trying templates which is not based on TuyaMCU as it is for a totally different hardware

In https://tasmota.github.io/docs/TuyaMCU/#fnid I can't seem to find a fnid especially targetted at controlling the leds
In Tuya doc https://developer.tuya.com/en/docs/iot/tuya-cloud-universal-serial-port-access-protocol?id=K9hhi0xxtn9cb, I can't seem neither tofind a command to control leds

It is likely that what you are trying to achieve is not possible on such Tuya device

You may want to return the unit and get the model that is listed in the template repository without Tuya
Move away from Tuya as much as possible as these devices don't give you much control
